client-config/src/main/java/com/palantir/conjure/java/client/config/ClientConfigurations.java

```diff
@@ -270,6 +298,59 @@ public String toString() {
         }
     }
 
+    private static final class NoProxyAwareSelector extends ProxySelector {
+        private static final ImmutableList<Proxy> NO_PROXIES = ImmutableList.of(Proxy.NO_PROXY);
+
+        private final ImmutableList<Proxy> proxies;
+        private final ImmutableSet<String> noProxyHosts; // host or domain suffixes
+
+        NoProxyAwareSelector(Proxy proxy, ImmutableSet<String> noProxyHosts) {
+            this.proxies = ImmutableList.of(proxy);
+            this.noProxyHosts = noProxyHosts;
+        }
+
+        @Override
+        public List<Proxy> select(URI uri) {
+            return shouldSkipProxy(uri) ? NO_PROXIES : proxies;
+        }
+
+        private boolean shouldSkipProxy(URI uri) {
```

Comment:
Thoughts on adding some debug/trace log lines to indicate when a proxy is used vs not? (can be useful to debug what's happening in certain networking setups)